### PR TITLE
ISTFT

### PR DESCRIPTION
**Pull request by jamarshon**
Inverse STFT (https://github.com/pytorch/audio/pull/130). Can't reopen previous one so here it is again

### CODE REVIEW

File: test/test_functional.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+class TestFunctional(unittest.TestCase):
+    data_sizes = [(2, 20), (3, 15)]
+    number_of_trials = 100
+    stored_rand_data = []
+    fixed_precision = int(1e10)
+
+    def setUp(self):
+        # we want to make sure that the random values are reproducible
+        self.stored_rand_data.clear()
+        torch.manual_seed(0)
```
Comment (cpuhrsch) [inline test/test_functional.py]: We still need to resolve this

File: test/test_functional.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+        self.assertTrue(sound.shape == estimate.shape, (sound.shape, estimate.shape))
+        self.assertTrue(torch.allclose(sound, estimate, atol=atol, rtol=rtol))
+
+    def _test_istft_is_inverse_of_stft(self, kwargs):
+        # generates a random sound signal for each tril and then does the stft/istft
+        # operation to check whether we can reconstruct signal
+        for i in range(len(self.data_sizes)):
+            for _ in range(self.number_of_trials):
+                sound = self._get_random_tensor(i)
```
Comment (cpuhrsch) [inline test/test_functional.py]: Each trial has the same data, right?

File: test/test_functional.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+import torchaudio
+import unittest
+import test.common_utils
+
+
+class TestFunctional(unittest.TestCase):
+    data_sizes = [(2, 20), (3, 15)]
+    number_of_trials = 100
+    stored_rand_data = []
+    fixed_precision = int(1e10)
```
Comment (cpuhrsch) [inline test/test_functional.py]: This should be a prime number

Comment (cpuhrsch) [reply test/test_functional.py]: If used in the random number generator

File: test/test_functional.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        self.stored_rand_data.clear()
+        torch.manual_seed(0)
+        for data_size in self.data_sizes:
+            rand_data1 = torch.randint(low=-self.fixed_precision, high=self.fixed_precision, size=data_size)
+            rand_data2 = torch.randint(low=-self.fixed_precision, high=self.fixed_precision, size=data_size)
+            self.stored_rand_data.append([rand_data1, rand_data2])
+
+    def _get_random_tensor(self, i):
+        # gets a random tensor of size data_sizes[i]. adds to previous tensors and then mods it.
+        rand_data = self.stored_rand_data[i]
```
Comment (cpuhrsch) [inline test/test_functional.py]: You don't need to store data for this. Take the first value (i), fix some constant to add (say c) and then some prime number (say p) to mod by.

So for element 0 you have i, for element 1 you have (i + c) mod  p for element 2 you have (element1 + c) mod p

File: test/test_functional.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
+import math
+
+import torch
+import torchaudio
+import unittest
+import test.common_utils
+
+
+class TestFunctional(unittest.TestCase):
+    data_sizes = [(2, 20), (3, 15)]
```
Comment (cpuhrsch) [inline test/test_functional.py]: We could add a few more and also degenerate cases like (1, 1) or (1, 0) etc.

File: test/test_functional.py [hunk 6]
```diff
@@ -9,25 +9,17 @@
...
     def setUp(self):
-        # we want to make sure that the random values are reproducible
-        self.stored_rand_data.clear()
-        torch.manual_seed(0)
+        self.random_tensor_generators.clear()
         for data_size in self.data_sizes:
-            rand_data1 = torch.randint(low=-self.fixed_precision, high=self.fixed_precision, size=data_size)
-            rand_data2 = torch.randint(low=-self.fixed_precision, high=self.fixed_precision, size=data_size)
-            self.stored_rand_data.append([rand_data1, rand_data2])
+            rtg = test.common_utils.RandomTensorGenerator(seed=0, size=data_size)
```
Comment (cpuhrsch) [inline test/test_functional.py]: Why do you create them here as opposed to on demand?

File: test/test_functional.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
+import math
+
+import torch
+import torchaudio
+import unittest
+import test.common_utils
+
+
+class TestFunctional(unittest.TestCase):
+    data_sizes = [(2, 20), (3, 15)]
```
Comment (cpuhrsch) [inline test/test_functional.py]: Please add more shapes here

File: test/test_functional.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        self.assertTrue(sound.shape == estimate.shape, (sound.shape, estimate.shape))
+        self.assertTrue(torch.allclose(sound, estimate, atol=atol, rtol=rtol))
+
+    def _test_istft_is_inverse_of_stft(self, kwargs):
+        # generates a random sound signal for each tril and then does the stft/istft
+        # operation to check whether we can reconstruct signal
+        for data_size in self.data_sizes:
+            rtg = test.common_utils.RandomTensorGenerator(seed=0, size=data_size)
+            for _ in range(self.number_of_trials):
+                sound = rtg.rand_float_tensor()
```
Comment (cpuhrsch) [inline test/test_functional.py]: This suffers from the same issue has using a regular number generator, which is that its stateful and the results dependent on the number of previous calls to rand_float_tensor

**Comment by vincentqb** [comment]
What are the differences with #130?

File: test/common_utils.py [hunk 9]
```diff
@@ -16,3 +17,34 @@ def create_temp_assets_dir():
...
+    c = 1
+    m = 2 ** 32
+
+    def __init__(self, seed, size):
+        # seed is an int and size is a tuple of ints
+        num_elements = 1
+        for s in size:
+            num_elements *= s
+
+        arr = [seed]
```
Comment (cpuhrsch) [inline test/common_utils.py]: seed is an int64_t if its a python integer, so it can't be the first number if you always mod by m for the other values.

File: test/common_utils.py [hunk 10]
```diff
@@ -16,3 +17,34 @@ def create_temp_assets_dir():
...
+        size (Tuple[int]): the size of the output tensor
+        a (int): the multiplier constant to the generator
+        c (int): the additive constant to the generator
+        m (int): the modulus constant to the generator
+    """
+    num_elements = 1
+    for s in size:
+        num_elements *= s
+
+    arr = [seed % m]
```
Comment (cpuhrsch) [inline test/common_utils.py]: this will give you the same return data for a whole range of seeds as opposed to using it within the first iteration of data generation.

Comment (cpuhrsch) [reply test/common_utils.py]: it can't be the first element of the generated data.

File: test/test_functional.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
+import math
+
+import torch
+import torchaudio
+import unittest
+import test.common_utils
+
+
+class TestFunctional(unittest.TestCase):
+    data_sizes = [(2, 20), (3, 15), (4, 10)]
```
Comment (cpuhrsch) [inline test/test_functional.py]: Please exercise (2, 0) or assert that an exception is raised.
